core/toolbox/toolbox.js

```diff
@@ -0,0 +1,944 @@
+ */
+Blockly.Toolbox.prototype.setSelectedItem = function(newItem) {
+  var oldItem = this.selectedItem_;
+
+  if ((!newItem && !oldItem) || (newItem && !newItem.isSelectable())) {
+    return;
+  }
+  newItem = /** @type {Blockly.SelectableToolboxItem} */ (newItem);
+
+  if (this.shouldDeSelectItem_(oldItem, newItem)) {
+    // Deselect the old item unless the old item is collapsible and has been
+    // previously clicked on.
+    this.selectedItem_ = null;
+    this.previouslySelectedItem_ = oldItem;
+    oldItem.setSelected(false);
+    Blockly.utils.aria.setState(/** @type {!Element} */ (this.contentsDiv_),
+        Blockly.utils.aria.State.ACTIVEDESCENDANT, '');
+  }
+
+  if (this.shouldSelectItem_(oldItem, newItem)) {
+    // Select the new item unless the old item equals the new item.
+    this.selectedItem_ = newItem;
+    this.previouslySelectedItem_ = oldItem;
+    newItem.setSelected(true);
+    Blockly.utils.aria.setState(/** @type {!Element} */ (this.contentsDiv_),
+        Blockly.utils.aria.State.ACTIVEDESCENDANT, newItem.getId());
+  }
+
+  this.updateFlyout_(oldItem, newItem);
+  this.fireSelectEvent_(oldItem, newItem);
+};
+
+/**
+ * Decides whether the old item should be deselected.
+ * @param {?Blockly.SelectableToolboxItem} oldItem The previously selected
+ *     toolbox item.
+ * @param {?Blockly.SelectableToolboxItem} newItem The newly selected toolbox
+ *     item.
+ * @return {boolean} True if the old item should be deselected, false otherwise.
+ * @protected
+ */
+Blockly.Toolbox.prototype.shouldDeSelectItem_ = function(oldItem, newItem) {
+  return !!(oldItem && (!oldItem.isCollapsible() || oldItem != newItem));
+};
+
+/**
+ * Decides whether the new item should be selected.
+ * @param {?Blockly.SelectableToolboxItem} oldItem The previously selected
+ *     toolbox item.
+ * @param {?Blockly.SelectableToolboxItem} newItem The newly selected toolbox
+ *     item.
+ * @return {boolean} True if the new item should be deselected, false otherwise.
+ * @protected
+ */
+Blockly.Toolbox.prototype.shouldSelectItem_ = function(oldItem, newItem) {
+  return !!(newItem && newItem != oldItem);
+};
+
+/**
+ * Selects the toolbox item by its position in the list of toolbox items.
+ * @param {number} position The position of the item to select.
+ * @public
+ */
+Blockly.Toolbox.prototype.selectItemByPosition = function(position) {
+  if (position > -1 && position < this.contents_.length) {
+    var item = this.contents_[position];
+    if (item.isSelectable()) {
+      this.setSelectedItem(item);
+    }
+  }
+};
+
+/**
+ * Decides whether to hide or show the flyout depending on the selected item.
+ * @param {?Blockly.SelectableToolboxItem} oldItem The previously selected toolbox item.
+ * @param {?Blockly.SelectableToolboxItem} newItem The newly selected toolbox item.
+ * @protected
+ */
+Blockly.Toolbox.prototype.updateFlyout_ = function(oldItem, newItem) {
+  if ((oldItem == newItem && !newItem.isCollapsible()) || !newItem ||
```

Comment:
It feels like `isCollapsible` is doing double duty. Sometimes it's about whether to do the accordion behaviour within the toolbox, and in other cases it's about whether to show contents in the flyout. Can we separate it into two state variables + getters?

Reply:
No longer an issue since collapsible category can have both blocks and categories, so now isCollapsible should only be in charge of the accordion functionality.